### PR TITLE
feat: Add happy path ACs for key rotation

### LINKED ISSUE #1315: 0063-VALK-validator_vega_master_keys.md
Add link to 0067 happy path AC for key rotation

Can we have additional coverage ......
AC parallel rotations in same block
AC rewards are received after rotation
AC staking delegation works after rotation

### PR DESCRIPTION
**Pull request by gordsport**
close #1315 .

@vega-paul regarding rewards and delegations should there be an AC that ensures that the balances are carried over or not?

### CODE REVIEW

File: protocol/0063-VALK-validator_vega_master_keys.md [hunk 1]
```diff
@@ -35,6 +35,10 @@ The public master key has to be added to validators' identities in the genesis c
 - A transaction submitted by an old hot-key after `target_block` is not associated by the network as being from a Validator (<a name="0063-VALK-003" href="#0063-VALK-003">0063-VALK-003</a>) 
 - Once `target_block` has been reached the network reports the new key as the validators hot-key (<a name="0063-VALK-004" href="#0063-VALK-004">0063-VALK-004</a>)
 - A key rotation submission which is not signed by the master key is rejected causing a transaction-error event that is visible by the whole network.  (<a name="0063-VALK-005" href="#0063-VALK-005">0063-VALK-005</a>)
+- It is possible to perform parallel key rotations rotations succesfully in the same block. (<a name="0063-VALK-006" href="#0063-VALK-006">0063-VALK-006</a>)
+- Once a validator master key has been rotated all applicable rewards are correctly received. (<a name="0063-VALK-007" href="#0063-VALK-007">0063-VALK-007</a>)
```
Comment (gordsport) [inline protocol/0063-VALK-validator_vega_master_keys.md]:
```suggestion
- Once a validator hot key has been rotated all applicable rewards are correctly received. (<a name="0063-VALK-007" href="#0063-VALK-007">0063-VALK-007</a>)
```

File: protocol/0063-VALK-validator_vega_master_keys.md [hunk 2]
```diff
@@ -35,6 +35,10 @@ The public master key has to be added to validators' identities in the genesis c
...
+- Once a validator master key has been rotated staking and delegation to the validator works as before. (<a name="0063-VALK-008" href="#0063-VALK-008">0063-VALK-008</a>)
```
Comment (gordsport) [inline protocol/0063-VALK-validator_vega_master_keys.md]:
```suggestion
- Once a validator hot key has been rotated staking and delegation to the validator works as before. (<a name="0063-VALK-008" href="#0063-VALK-008">0063-VALK-008</a>)
```

File: protocol/0063-VALK-validator_vega_master_keys.md [hunk 3]
```diff
@@ -35,6 +35,10 @@ The public master key has to be added to validators' identities in the genesis c
...
+- - Once a validator master key has been rotated self-staking and self-delegation for the validator works as before. (<a name="0063-VALK-009" href="#0063-VALK-009">0063-VALK-009</a>)
```
Comment (gordsport) [inline protocol/0063-VALK-validator_vega_master_keys.md]:
```suggestion
- Once a validator hot key has been rotated self-staking and self-delegation for the validator works as before. (<a name="0063-VALK-009" href="#0063-VALK-009">0063-VALK-009</a>)
```

**Review by Vegaklaus — COMMENTED** [review]
I'd replace 'rotated' with 'rotated and activated' to be clear.

**Comment by Vegaklaus** [comment]
We have an undefined behaviour if a validator changes the hot key again before the previous one is invalidated. My interpretation is that the old key to be deactivated is the one with the preceding key number.

Also, do we reject wrong key numbers at this point ? This would need an AC
"If a validator submits a key-rotate with an invalid key number, it is rejected. A key number is invalid if it is not 'last active key'+1"

**Comment by wwestgarth** [comment]
The current core implementation doesn't allow multiple "pending" key rotations at the same time (we return a tx-error if you try to submit another on), so just an FYI that there will be a little core work for those ACs.

And just to confirm, the idea for `0063-VALK-009` is that you can submit another key rotation only if the second one is to become activate at a higher block than the first, and its key sequence number is higher? i.e you can't cancel/bring forward a key rotation by submitting a new one for an earlier?

**Comment by Vegaklaus** [comment]
> The current core implementation doesn't allow multiple "pending" key rotations at the same time (we return a tx-error if you try to submit another on), so just an FYI that there will be a little core work for those ACs.
> 
> And just to confirm, the idea for `0063-VALK-009` is that you can submit another key rotation only if the second one is to become activate at a higher block than the first, and its key sequence number is higher? i.e you can't cancel/bring forward a key rotation by submitting a new one for an earlier?

I fixed the AC to allow both solutions; so it's now legal to either reject the additional transactions or manage multiple ones appropriately.

**Review by Vegaklaus — DISMISSED** [review]

**Review by Vegaklaus — APPROVED** [review]

**Pull request merged**